### DISCUSSION TITLE
Upgrade mantis java version to JDK 17

### DISCUSSION THREAD

**Discussion by garryliu-stripe** [Netflix/mantis#705] [General]
Hey all! (probably @Andyz26  and @sundargates). What do you guys think of upgrading JDK to 17? I found an [old issue](https://github.com/Netflix/mantis/issues/355) from March 2023 regarding this. There was a [attempt](https://github.com/Netflix/mantis/pull/390) by @Adithya2907 and the attempt builds.

If we agree on kicking off the upgrade, we(Stripe) are happy to help with compatibility tests.

**Comment by Andyz26** [comment]
hi @garryliu-stripe thanks for looking into this. I think the main components can be upgraded into jdk17+ safely except the mantis publish related modules and the mantis client (essentially any libraries that could be used by other applications which may still require jdk8 support).

**Comment by codyrioux** [comment]
A few small updates on this end:

* [[Draft] Upgrade JDK target to 17](https://github.com/Netflix/mantis/pull/735) compiles and all tests pass on an M1 Max Mac with JDK 17.
* [[Draft] Upgrade JDK target to 17](https://github.com/Netflix/mantis/pull/735) compiles and all tests pass (except container tests, I don't have docker) on Ubuntu 22.04 on Windows Subsystem for Linux with JDK 17.
* I've done a code review for the [pull request](https://github.com/Netflix/mantis/pull/735) and besides some minor nitpicks it looks perfect.
* I have cut a release candidate off the branch as `4.0.0-rc.1` for testing internally at Stripe, but am no long at Stripe so I cannot test myself.

@andresgalindo-stripe Would love to hear that it at least all fires up in `mantisdev`.

**Comment by Andyz26** [comment]
[update] 
@andresgalindo-stripe @kmg-stripe I am going to merge the JDK 17 upgrade into master branch and update the CI requirements on the master branch rule. At the mean time i've created a branch releases/3.x to be used as a hot fix branch if needed.

**Reply by kmg-stripe** [reply]
@Andyz26 ack, ty!  we'll continue to work off the 3.x tags.  we are scoping out JDK 17 on our end.  it is currently unclear when we will upgrade to 17.

**Reply by Andyz26** [reply]
@kmg-stripe since the master branch is merged now with JDK17 do you plan to back port the changes from master to 3.x release branch or the other way around?

**Comment by kmg-stripe** [comment]
@Andyz26 - the jdk 17 upgrade has been really stop-and-go, and we haven't pulled in upstream for a while.  to speed up pulling fixes in, we are thinking of back porting to a branch, maybe a `2.x` branch.  any concerns?

cc: @andresgalindo-stripe @mcowgill-stripe

**Comment by Andyz26** [comment]
yeah i am fine with that for now. Fyi we are planning a major refactor into
the scheduling and resource cluster integration in next few months so
cherry-picking individual changes might get trickier and moving back to
master branch on your side would be a better option long term still.

**Reply by kmg-stripe** [reply]
makes sense, thanks!  we will prioritize upgrading to 17 after we pull in from upstream.  we have been bit by the split brain issue a few times, so want to get that fix in soon.

**Comment by kmg-stripe** [comment]
Hey @Andyz26 - Sorry to keep pestering you, but we were planning to backport the cherry-picked changed in https://github.com/Netflix/mantis/pull/805 to `3.x` tomorrow 11/6.  Let us know if you have any concerns.  Thanks!

**Comment by Andyz26** [comment]
sorry about the late reply i've been on vacation. Yeah we've moved away
from the 3.x branch so feel free to backport changes there.
